KEYCLOAK-13698 - SAML Client - Add certificate info to signature

lscorcia opened the pull request:
Adds the X509Data tag to the XML Document signature in SAML AuthnRequests

hmlnarik reviewed (changes requested):
Thank you for the PR.

Could you please also provide the tests? See e.g. org.keycloak.testsuite.broker.KcSamlSignedBrokerTest.

lscorcia commented:
Hi @hmlnarik, thanks for your review and sorry for the delay in my reply.

I have been trying to understand how to add the appropriate tests to `org.keycloak.testsuite.broker.KcSamlSignedBrokerTest`, but I fail to see anything that analyzes the emitted login request before it is sent to the remote IdP. This class seems to verify the response more than the requests.

I spent some time looking around and I noticed 
https://github.com/keycloak/keycloak/blob/020ffd37ee70f9a65762e1454211ab65afedf3a0/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/SamlClient.java#L161-L168
and
https://github.com/keycloak/keycloak/blob/020ffd37ee70f9a65762e1454211ab65afedf3a0/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/SamlClient.java#L278-L284
where the tests sign the request without supplying the certificate info just like in the code I proposed to amend, but there is no clear certificate parameter around those lines.

Am I completely off track here? I have to say I feel a bit lost.

lscorcia commented:
Rebased to current, updated all tests that only used KeyPairs to sign the documents and added a test to explicitly verify that the X509Data element is present in the generated AuthnRequest.

The test failure in the CI build doesn't seem to be related to my changes.

It took me a while to understand how to create a functional test, I hope that's what was required for acceptance, otherwise please let me know.

hmlnarik commented on testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/util/saml/CreateAuthnRequestStepBuilder.java:
```diff
@@ -79,9 +79,10 @@ public CreateAuthnRequestStepBuilder relayState(String relayState) {
     }
 
-    public CreateAuthnRequestStepBuilder signWith(String signingPrivateKeyPem, String signingPublicKeyPem) {
+    public CreateAuthnRequestStepBuilder signWith(String signingPrivateKeyPem, String signingPublicKeyPem, String signingCertificate) {
```
Changing this method signature leads to many unnecessary changes. There is no need to do so, please revert, as well as related changes in the testsuite. Feel free to create a overloaded version of this method with this signature, but retain the previous version.

lscorcia replied:
Thanks for your review! I reverted the unnecessary changes you pointed out and pushed two separate commits, the first with the bare minimum issue fix, the related test and the correct commit message, and a second one that adds the ability to specify the certificate to be used in AuthnRequest/LogoutRequest tests.

Please let me know if they are both good and I'll squash the commits, but if you prefer not touching the other test files I will just revert the second one.

hmlnarik reviewed (changes requested):
Apologies for the delay in PR reviews. Thank you, the change makes sense. The only two requirements before merging are:
1. To keep the PR minimal and free of unnecessary changes.
2. To update commit message to start with KEYCLOAK-xyz (remove the "Fix " prefix)

lscorcia committed:
KEYCLOAK-13698 - Add certificate info to signature

Adds the X509Data tag to the XML Document signature in AuthnRequests

lscorcia committed:
Allow Authn/Logout request tests to specify the signing certificate

hmlnarik reviewed (approved):
Thanks, the changes looks good to me now!
